File: C++/enumerations/permutations_and_subsets.cpp
#include <vector>
#include <string>
#include <iostream>
#define BLOCK 1000000000 // значение, которого нет в массиве

void genAllPermutations(std::vector<int> arr, std::vector<int> v, int n){
    if (v.size() == n){
        for (int el: v) printf("%d ", el);
        printf("\n");
    }
    else{
        for (int i = 0; i < arr.size(); i++){
            if (arr[i] != BLOCK){
                std::vector<int> new_arr = arr;
                new_arr[i] = BLOCK;
                v.push_back(arr[i]);
                genAllPermutations(new_arr, v, n);
                v.pop_back();
            }
        }
    }
}

void genAllSubsets(std::vector<int> arr, std::vector<int> v, int n){
    if (v.size() == n){
        for (int el: v) printf("%d ", el);
        printf("\n");
    }
    else{
        for (int i = 0; i < arr.size(); i++){
            if (arr[i] != BLOCK){
                v.push_back(arr[i]);
                arr[i] = BLOCK;
                genAllSubsets(arr, v, n);
                v.pop_back();
            }
        }
    }
}

int main(){
    int len;
    std::cin >> len;
    std::vector<int> arr(len);
    for (int i = 0; i < len; i++){
        std::cin >> arr[i];
    }
    for (int n = 1; n <= len; n++){
        genAllSubsets(arr, std::vector<int>{}, n);
    }
    return 0;
}

File: Python/enumerations/permutations_and_subsets.py
def gen_all_permutations(arr: list[...], v: list[int], n: int) -> None:
    if len(v) == n:
        print(*v, sep='')
    else:
        for i in range(len(arr)):
            if arr[i] is not None:
                new_arr = arr.copy()
                new_arr[i] = None
                gen_all_permutations(new_arr, v + [arr[i]], n)

def gen_all_subsets(narr: list[...], v: list[int], n: int) -> None:
    arr = narr.copy()
    if len(v) == n:
        print(*v, sep='')
    else:
        for i in range(len(arr)):
            if arr[i] is not None:
                el = arr[i]
                arr[i] = None
                gen_all_subsets(arr, v + [el], n)

array = list(map(int, input().split()))
for length in range(1, len(array) + 1):
    gen_all_subsets(array, [], length)